wit/tcp.wit

```diff
@@ -185,21 +200,69 @@ interface tcp {
 	keep-alive: func(this: tcp-socket) -> result<bool, error-code>
 	set-keep-alive: func(this: tcp-socket, value: bool) -> result<_, error-code>
 
+	/// Equivalent to the TCP_KEEPIDLE socket option.
+	/// 
+	/// # Typical errors
+	/// - `concurrency-conflict`: (set) A `bind`, `connect` or `listen` operation is already in progress. (EALREADY)
+	keep-alive-idle-time: func(this: tcp-socket) -> result<duration, error-code>
+	set-keep-alive-idle-time: func(this: tcp-socket, value: duration) -> result<_, error-code>
+
+	/// Equivalent to the TCP_KEEPINTVL socket option.
+	/// 
+	/// # Typical errors
+	/// - `concurrency-conflict`: (set) A `bind`, `connect` or `listen` operation is already in progress. (EALREADY)
+	keep-alive-interval: func(this: tcp-socket) -> result<duration, error-code>
+	set-keep-alive-interval: func(this: tcp-socket, value: duration) -> result<_, error-code>
+
+	/// Equivalent to the TCP_KEEPCNT socket option.
+	/// 
+	/// # Typical errors
+	/// - `concurrency-conflict`: (set) A `bind`, `connect` or `listen` operation is already in progress. (EALREADY)
+	keep-alive-count: func(this: tcp-socket) -> result<u32, error-code>
+	set-keep-alive-count: func(this: tcp-socket, value: u32) -> result<_, error-code>
+
 	/// Equivalent to the TCP_NODELAY socket option.
 	/// 
 	/// # Typical errors
 	/// - `concurrency-conflict`: (set) A `bind`, `connect` or `listen` operation is already in progress. (EALREADY)
 	no-delay: func(this: tcp-socket) -> result<bool, error-code>
 	set-no-delay: func(this: tcp-socket, value: bool) -> result<_, error-code>
+
+	/// Equivalent to the TCP_CORK/TCP_NOPUSH socket option.
+	/// 
+	/// # Typical errors
+	/// - `concurrency-conflict`: (set) A `bind`, `connect` or `listen` operation is already in progress. (EALREADY)
+	no-push: func(this: tcp-socket) -> result<bool, error-code>
+	set-no-push: func(this: tcp-socket, value: bool) -> result<_, error-code>
```

Comment:
> Could we be confident that TCP_CORK and TCP_NOPUSH are the same?

Do we need to? I consider `no-push` to be an optimization hint, telling the host "Hey, I'm about to make many consecutive Write calls. Try to optimize for network efficiency instead of latency and feel free to optimize the TCP frames accordingly". Both Linux' TCP_CORK and BSD's TCP_NOPUSH fit that description. Minor implementation differences are to be expected. For example, Windows doesn't support it at all, but can fall back to regular non-NODELAY behavior (or just error out). In the end, exactly the same bytes will be received on the other end, either way.

The blog post you referenced mentions a difference in what happens when the cork ends. Both implementations are equally "correct" in my mind. If an application wanted the uncork to also immediately send out the data without delay  (Linux' behavior), then the application should have enabled `no-delay`.

Reply:
Would you mind adding a section to the documentation to discuss issues like this, and mapping `IP_DONTFRAG` to `IP_MTU_DISCOVER`, and other portability concerns? The Notes in the table are good, but it'd be helpful to have more detail somewhere. My experience here as someone familiar with sockets but has never had to know what eg. `TCP_CORK` does is that the implementation work shifts from "wiring things up" to "scouring through documentation and learning all about TCP protocol flags and kernel buffering policies and interactions with other obscure socket options to figure out if I need to do anything special to wire things up".

Reply:
I've added a couple of `# Implementors note`  sections in the new PRs